modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningMetadataExtractor.java

```diff
@@ -293,16 +304,18 @@ private void extractFromValues(
 
             List<RexNode> andNodes = new ArrayList<>(keysList.size());
             for (int key : keysList) {
+                RexNode lit = items.get(mapping != null ? mapping.getSourceOpt(key) : key);
```

Comment:
Why `mapping.getSourceOpt(key)` can never produce `-1` ?

Reply:
i radically refactor this place, can you take a look once more plz ?